### ISSUE TITLE
tls metrics is not reliable

### ISSUE THREAD

**Issue by BusyJay** [tikv/tikv#12562]
### What version of TiKV are you using?
Nightly

### What operating system and CPU are you using?
Linux

### Steps to reproduce
Run test case `raftstore::test_stats::test_query_stats` several times

### What did you expect?
Test case always pass.

### What did happened?
It will panic eventually.

tls metrics relies on periodically calling `tls_flush` to aggregate counters. The metrics is also used for flow reports and can affect the PD scheduling.

The fundamental pitfall of tls is that it's maintained by thread itself, other threads have no access to the storage, hence it's critical to ensure all threads that have mutation must call `tls_flush` afterwards, otherwise the metrics will be missed. Coprocessor thread pool and scheduler thread pool are hooked up. However, there is no such guarantee in grpc thread.

And because we are using async/await, which thread the execution takes place is unpredictable, so the metrics can be mutated in all possible threads. So the metrics collected in grpc thread will be missed and cause the case panic.

To get around the problem, we can also add `tls_flush` to all grpc threads.

To fix the fundamental problems, we may need to redesign and reimplement TLS. For example, instead of using thread local variables, we can use global metrics and shard one for each threads.

```
|<pading> metrics <pading>|<pading> metrics <pading>| ...
```

And a thread can be used for polling and aggregating metrics periodically.

Another alternatives is using per-core metrics just like RocksDB does. http://rocksdb.org/blog/2017/05/14/core-local-stats.html

Previous discussions #2651 #4565

/cc @breezewish @lhy1024

**Comment by breezewish** [comment]
I agree that the current metrics design is too simple. The usage is also becoming more complicated with comes with "static metrics". There could be a refinement for the rust-prometheus crate for supporting such demands.

About sharing the metrics ref, I think we need to discuss about this further: What if a thread is created and destroyed? How could we maintain the lifecycle of the global ref correctly?

The first idea comes up to me is to have something like pthread ThreadID, that could be used to know a thread is destroyed externally.

**Comment by BusyJay** [comment]
Using per core metrics doesn't require tracking thread lifetime.

**Comment by breezewish** [comment]
> Using per core metrics doesn't require tracking thread lifetime.

One thing worries about me is how to know which core the call is working on *efficiently*. In Top SQL we use `RDTSCP` instruction which contains a CPU core information. However it is not portable.
